build-plugins/build-support/src/main/kotlin/aws/sdk/kotlin/gradle/plugins/artifactsizemetrics/CollectDelegatedArtifactSizeMetrics.kt

```diff
@@ -31,79 +34,52 @@ internal abstract class CollectDelegatedArtifactSizeMetrics : DefaultTask() {
 
-        val artifactSizeMetricsFileKeys = getFileKeys() ?: throw AwsSdkGradleException("Unable to list objects from artifact size metrics bucket")
-
-        val pluginConfig = this.project.rootProject.extensions.getByType(ArtifactSizeMetricsPluginConfig::class.java)
-
-        val relevantArtifactSizeMetricsFileKeys = artifactSizeMetricsFileKeys.filter {
-            it?.startsWith("[TEMP]${pluginConfig.projectRepositoryName}-$identifier-") == true
-        }
-
-        val artifactSizeMetricsFiles = getFiles(relevantArtifactSizeMetricsFileKeys)
+        val artifactSizeMetricsFileKeys = getFileKeys(identifier) ?: throw AwsSdkGradleException("Unable to list objects from artifact size metrics bucket")
+        val artifactSizeMetricsFiles = getFiles(artifactSizeMetricsFileKeys)
         val combined = combine(artifactSizeMetricsFiles)
 
         metricsFile.asFile.get().writeText(combined)
     }
 
-    private fun getFileKeys(): List<String?>? = runBlocking {
+    private fun getFileKeys(identifier: String): List<String>? = runBlocking {
         S3Client.fromEnvironment().use { s3 ->
             return@runBlocking s3.listObjects {
                 bucket = S3_ARTIFACT_SIZE_METRICS_BUCKET
-            }.contents?.map { it.key }
+                prefix = pluginConfig.bucketPrefixOverride ?: "[TEMP]${pluginConfig.projectRepositoryName}-$identifier-"
+            }.contents?.map {
+                it.key ?: throw AwsSdkGradleException("A file from the artifact size metrics bucket is missing a key")
+            }
         }
     }
 
-    private fun getFiles(keys: List<String?>): List<String> {
-        val files = mutableListOf<String>()
+    private fun getFiles(keys: List<String>): List<String> = runBlocking {
+        val files = mutableListOf<Deferred<String>>()
 
-        runBlocking {
-            S3Client.fromEnvironment().use { s3 ->
-                keys.forEach { k ->
-                    k?.let {
+        S3Client.fromEnvironment().use { s3 ->
+            keys.forEach { k ->
+                files.add(
+                    async {
                         s3.getObject(
                             GetObjectRequest {
                                 bucket = S3_ARTIFACT_SIZE_METRICS_BUCKET
                                 key = k
                             },
                         ) { file ->
-                            files.add(
-                                file.body?.decodeToString() ?: throw AwsSdkGradleException("Metrics file $file is missing a body"),
-                            )
+                            file.body?.decodeToString() ?: throw AwsSdkGradleException("Metrics file $k is missing a body")
                         }
-                    }
-                }
+                    },
+                )
             }
+            return@runBlocking files.awaitAll()
         }
-
-        return files
     }
```

Comment:
**Nit**: You don't need a mutable list reference to track the deferred results, you can just use `map`:

```kotlin
private fun getFiles(keys: List<String>) =
    S3Client.fromEnvironment().use { s3 ->
        keys
            .map { k ->
                async {
                    s3.getObject(
                        GetObjectRequest {
                            bucket = S3_ARTIFACT_SIZE_METRICS_BUCKET
                            key = k
                        },
                    ) { it.body?.decodeToString() ?: throw AwsSdkGradleException("Metrics file $k is missing a body") }
                }
            }.awaitAll()
    }
```

For clarity you can even extract the `getObject` call to a helper function:

```kotlin
private suspend fun S3Client.getObjectAsText(key: String) =
    getObject(
        GetObjectRequest {
            bucket = S3_ARTIFACT_SIZE_METRICS_BUCKET
            key = k
        },
    ) { it.body?.decodeToString() ?: throw AwsSdkGradleException("Metrics file $file is missing a body") }

private fun getFiles(keys: List<String>) =
    S3Client.fromEnvironment().use { s3 ->
        keys.map { k ->
            async { s3.getObjectAsText(k) }
        }.awaitAll()
    }
```